commons/src/scanner/symbol/iterator/mod.rs

```diff
@@ -262,13 +300,16 @@ impl<'input> Iterator for SymbolIterator<'input> {
     type Item = &'input Symbol<'input>;
 
     fn next(&mut self) -> Option<Self::Item> {
-        if self.end_reached() {
+        if self.prefix_mismatch || self.end_reached() {
             return None;
         }
 
         if let Some(end_fn) = self.end_match.clone() {
+            self.next_matching = true;
+
             if (end_fn)(self) {
                 self.iter_end = true;
+                self.next_matching = false;
```

Comment:
I updated the doc on `next_matching`. Hopefully this makes this code clearer.

In short: Setting `next_matching` flag in `next()` prevents matching in `peeking_next()` during execution of the matching functions in `next()`. Matching functions must get the symbols as is, because they define what symbols must be skipped.